src/Approximations/decompositions.jl

```diff
@@ -439,3 +439,150 @@ The template direction approximation of the projection of `S`.
+```
+
+Projecting along dimension `2` only:
+
+```jldoctest project_halfspace
+julia> project(H, [1, 2])
+HalfSpace{Float64}([1.0, 1.0], 1.0)
+```
+In general, use the call syntax `project(H, constrained_dimensions(H))` to return
+the half-space projects in the dimensions where it is constrained only:
+
+```jldoctest project_halfspace
+julia> project(H, constrained_dimensions(H))
+HalfSpace{Float64}([1.0, 1.0], 1.0)
+```
+"""
+function project(H::HalfSpace{N}, block::AbstractVector{Int}) where {N}
+    if constrained_dimensions(H) ⊆ block
+        return HalfSpace(H.a[block], H.b)
+    else
+        error("the concrete projection of a half-space " *
+              "for a general block structure is not implemented yet")
+    end
+end
+
+"""
+    project(P::HPolyhedron{N}, block::AbstractVector{Int}) where {N}
+
+Concrete projection of a polyhedron in half-space representation.
+
+### Input
+
+- `P`        -- set
+- `block`    -- block structure, a vector with the dimensions of interest
+- `set_type` -- (optional, default `HPolyhedron`) used for dispatch
+
+### Output
+
+A set representing the projection of `P` on the dimensions specified by `block`.
+
+### Notes
+
+Currently only the case where the unconstrained dimensions of `P` are a subset
+of the `block` variables is implemented.
+
+### Examples
+
+Consider the four-dimensional cross-polytope (unit ball in the 1-norm):
+
+```jldoctest project_hpolyhedron
+julia> using LazySets.Approximations: project
+
+julia> P = convert(HPolyhedron, Ball1(zeros(4), 1.0));
+```
+
+All dimensions are constrained, and computing the (trivial) projection on the whole
+space behaves as expected:
+
+```jldoctest project_hpolyhedron
+julia> constrained_dimensions(P)
+4-element Array{Int64,1}:
+ 1
+ 2
+ 3
+ 4
+
+julia> P_1234 = project(P, [1, 2, 3, 4]);
+
+julia> P_1234 == P
+true
+```
+Each constraint of the cross polytope is constrained in all dimensions.
+
+Now let's take a ball in the infinity norm and reomove some constraints:
+
+```jldoctest project_hpolyhedron
+julia> B = BallInf(zeros(4), 1.0);
+
+julia> c = constraints_list(B)[1:2]
+2-element Array{HalfSpace{Float64},1}:
+ HalfSpace{Float64}([1.0, 0.0, 0.0, 0.0], 1.0)
+ HalfSpace{Float64}([0.0, 1.0, 0.0, 0.0], 1.0)
+
+julia> P = HPolyhedron(c);
+
+julia> constrained_dimensions(P)
+2-element Array{Int64,1}:
+ 1
+ 2
+```
+
+Finally we take the concrete projection onto variables `1` and `2`:
+
+```jldoctest project_hpolyhedron
+julia> project(P, [1, 2]) |> constraints_list
+2-element Array{HalfSpace{Float64},1}:
+ HalfSpace{Float64}([1.0, 0.0], 1.0)
+ HalfSpace{Float64}([0.0, 1.0], 1.0)
+```
+"""
+function project(P::HPolyhedron{N}, block::AbstractVector{Int}) where {N}
+    if constrained_dimensions(P) ⊆ block
+        constraints = Vector{HalfSpace{N}}(undef, length(constraints_list(P)))
+        @inbounds for (i, ci) in enumerate(constraints_list(P))
+            constraints[i] = HalfSpace(ci.a[block], ci.b)
+        end
```

Comment:
I would use list comprehension here (after all, I learned it from you):
```julia
constraints = [HalfSpace(c.a[block], c.b) for c in constraints_list(P)]
```

Reply:
yes, that works as well and list comprehension is more readable and i prefer it! 😄 

but we know it allocates, so it is more efficient to preallocating the output. so i would avoid it in code that be performance critical (i think that `project` could be such a case). Let's see:

```julia
julia> using Revise, LazySets

julia> P2 = HPolyhedron([rand(HalfSpace, dim=2) for _ in 1:2]);

julia> P100 = HPolyhedron([rand(HalfSpace, dim=2) for _ in 1:100]);


julia> P100_high = HPolyhedron([rand(HalfSpace, dim=100) for _ in 1:100]);

julia> using LazySets.Approximations: decompose

julia> using LazySets.Approximations: project

julia> using LazySets.Approximations: project_list_comprehension

julia> block2 = [1,2];

julia> block100 = 1:100;

julia> @btime project($P2, $block2)
  850.766 ns (19 allocations: 976 bytes)
HPolyhedron{Float64}(HalfSpace{Float64}[HalfSpace{Float64}([-0.651225, -1.93049], -0.384613), HalfSpace{Float64}([0.87216, -0.352526], 1.03974)])

julia> @btime project_list_comprehension($P2, $block2)
  1.284 μs (25 allocations: 1.25 KiB)
HPolyhedron{Float64}(HalfSpace{Float64}[HalfSpace{Float64}([-0.651225, -1.93049], -0.384613), HalfSpace{Float64}([0.87216, -0.352526], 1.03974)])

julia> @btime project($P100, $block2);
  26.949 μs (705 allocations: 32.36 KiB)

julia> @btime project_list_comprehension($P100, $block2);
  37.083 μs (907 allocations: 44.91 KiB)

julia> @btime project($P100_high, $block100);
  114.142 μs (721 allocations: 194.85 KiB)

julia> @btime project_list_comprehension($P100_high, $block100);
  453.540 μs (2023 allocations: 933.18 KiB)
```

Reply:
And what about the alternative below?

Reply:
the one that i timed is `project_list_comprehension` from the code snippet in [this comment](https://github.com/JuliaReach/LazySets.jl/pull/1213#discussion_r263501143). it is the same function as in [this other comment](https://github.com/JuliaReach/LazySets.jl/pull/1213#discussion_r263494571)

Reply:
And what about this?
```julia
function project(P::HPolyhedron{N}, block::AbstractVector{Int}) where {N}
    constraints = Vector{HalfSpace{N}}(undef, length(constraints_list(P)))
    @inbounds for (i, ci) in enumerate(constraints_list(P))
            constraints[i] = project(ci, block)
    end
    return HPolyhedron(constraints)
end
```

Reply:
let

```julia
function project_list_comprehension_v2(P::HPolyhedron{N}, block::AbstractVector{Int}) where {N}
    if constrained_dimensions(P) ⊆ block
        return HPolyhedron([HalfSpace(c.a[block], c.b) for c in constraints_list(P)])
    else
        error("the concrete projection of a half-space " *
              "for a general block structure is not implemented yet")
    end
end
```
then:

```julia
julia> using Revise, LazySets
[ Info: Recompiling stale cache file /Users/forets/.julia/compiled/v1.1/LazySets/NjrGc.ji for LazySets [b4f0291d-fe17-52bc-9479-3d1a343d9043]

julia> P2 = HPolyhedron([rand(HalfSpace, dim=2) for _ in 1:2]);

julia> P100 = HPolyhedron([rand(HalfSpace, dim=2) for _ in 1:100]);

julia> P100_high = HPolyhedron([rand(HalfSpace, dim=100) for _ in 1:100]);

julia> using LazySets.Approximations: project

julia> using LazySets.Approximations: project_list_comprehension

julia> using LazySets.Approximations: project_list_comprehension_v2

julia> block2 = [1,2];

julia> block100 = 1:100;

julia> @btime project($P2, $block2)
  836.288 ns (19 allocations: 976 bytes)
HPolyhedron{Float64}(HalfSpace{Float64}[HalfSpace{Float64}([-0.472929, -0.672941], -1.27841), HalfSpace{Float64}([2.76533, -0.389741], 0.0295013)])

julia> @btime project_list_comprehension($P2, $block2)
  1.203 μs (25 allocations: 1.25 KiB)
HPolyhedron{Float64}(HalfSpace{Float64}[HalfSpace{Float64}([-0.472929, -0.672941], -1.27841), HalfSpace{Float64}([2.76533, -0.389741], 0.0295013)])

julia> @btime project_list_comprehension_v2($P2, $block2);
  974.917 ns (21 allocations: 1.00 KiB)

julia> @btime project($P100, $block2);
  26.661 μs (705 allocations: 32.36 KiB)

julia> @btime project_list_comprehension($P100, $block2);
  37.338 μs (907 allocations: 44.91 KiB)

julia> @btime project_list_comprehension_v2($P100, $block2);
  25.706 μs (707 allocations: 32.41 KiB)

julia> @btime project($P100_high, $block100);
  112.022 μs (721 allocations: 196.41 KiB)

julia> @btime project_list_comprehension($P100_high, $block100);
  447.615 μs (2022 allocations: 934.73 KiB)

julia> @btime project_list_comprehension_v2($P100_high, $block100);
  107.990 μs (722 allocations: 196.45 KiB)
```

i am tempted to using `project_list_comprehension_v2`  now, since it as efficient as unrolling the for loop but it is more readable 👍

Reply:
Isn't this what I proposed at the beginning? :blush:

Reply:
the first alternative that i tried was `project_list_comprehension` as in the bottom of [this comment](https://github.com/JuliaReach/LazySets.jl/pull/1213#discussion_r263501143). then i tried the other version `project_list_comprehension_v2` which just doesn't call `project(...)`. i wonder why it has this performance penalty. maybe if we need to add `@inline`?